### PR TITLE
add mashmap - long read optimised multispecies alignments - produces paf

### PR DESCRIPTION
**Pull request by fubar2**
FOR CONTRIBUTOR:
* [X] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [X] - License permits unrestricted use (educational + commercial)
* [X] - This PR adds a new tool or tool collection

https://github.com/marbl/MashMap

MashMap implements a fast and approximate algorithm for computing local alignment boundaries between long DNA sequences. It can be useful for mapping genome assembly or long reads (PacBio/ONT) to reference genome(s).

### CODE REVIEW

File: tools/mashmap/mashmap.xml [hunk 1]
```diff
@@ -0,0 +1,88 @@
+<tool name="mashmap" id="mashmap" version="3.1.3" profile="22.05">
+  <!--Source in git at: https://github.com/fubar2/galaxy_tf_overlay-->
+  <!--Created by toolfactory@galaxy.org at 23/02/2024 21:27:46 using the Galaxy Tool Factory.-->
+  <description>Fast local alignment boundaries</description>
+  <requirements>
+    <requirement version="3.1.3" type="package">mashmap</requirement>
+  </requirements>
+  <version_command><![CDATA[echo "3.1.3"]]></version_command>
```
Comment (bgruening) [inline tools/mashmap/mashmap.xml:8]: is there a `mashmap --version` command?

File: tools/mashmap/mashmap.xml [hunk 2]
```diff
@@ -0,0 +1,88 @@
+<tool name="mashmap" id="mashmap" version="3.1.3" profile="22.05">
```
Comment (bgruening) [inline tools/mashmap/mashmap.xml:1]: can you please use a macro here ... https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#tool-versions

File: tools/mashmap/mashmap.xml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+mashmap --pi '$perc_identity' -s '$seqLength' -f '$filtermode' $dense \
+#if int($sketchSize) > 0:
+-J '$sketchSize' \
+#end if
+#if len($reflist) == 1:
+ -r '$reflist' -q '$query' && 
+#else
+--rl 'reflist' -q '$query' && 
+#end if
+cp 'mashmap.out'  $mashout]]></configfile>
```
Comment (bgruening) [inline tools/mashmap/mashmap.xml]:
```suggestion
mv mashmap.out  '$mashout']]></configfile>
```

File: tools/mashmap/mashmap.xml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ 
+ Mashmap computes alignment boundaries and identity estimates using k-mers. It does not compute the alignments explicitly, but rather estimates an unbiased k-mer based Jaccard similarity using a combination of minmers (a novel winnowing scheme) and MinHash. This is then converted to an estimate of sequence identity using the Mash distance. An appropriate k-mer sampling rate is automatically determined using the given minimum local alignment length and identity thresholds.
+ 
+ As an example, Mashmap can map a human genome assembly to the human reference genome in about one minute total execution time and < 4 GB memory using just 8 CPU threads, achieving more than an order of magnitude improvement in both runtime and memory over alternative methods. We describe the algorithms associated with Mashmap, and report on speed, scalability, and accuracy of the software in the publications listed below. Unlike traditional mappers, MashMap does not compute exact sequence alignments. In future, we plan to add an optional alignment support to generate base-to-base alignments.
+ 
+ Map set of query sequences against a reference genome:
+ 
+ mashmap -r reference.fna -q query.fa
```
Comment (bgruening) [inline tools/mashmap/mashmap.xml:71]:
```suggestion
```

File: tools/mashmap/mashmap.xml [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+ 
+ The output is space-delimited with each line consisting of query name, length, 0-based start, end, strand, target name, length, start, end and mapping nucleotide identity.
+ 
+ Map set of query seqences against a list of reference genomes:
+ 
+ mashmap --rl referenceList.txt -q query.fa
```
Comment (bgruening) [inline tools/mashmap/mashmap.xml:77]:
```suggestion
```

File: tools/mashmap/mashmap.bash.txt [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#raw
```
[the remark this answers is not in the document]
Comment (fubar2) [reply tools/mashmap/mashmap.bash.txt:1]: sure. It's useful for recreating the tool and although I would prefer to keep it as an artefact, I will remove if you wish.

File: tools/mashmap/mashmap.xml [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+      <option value="--dense">Dense sketching</option>
+    </param>
+    <param name="filtermode" type="select" label="Filter mode" help="Mashmap implements a plane-sweep based algorithm to perform the alignment filtering. Similar to delta-filter in nucmer, different filtering options are provided that are suitable for long read or assembly mapping. Option -f map is suitable for reporting the best mappings for long reads, whereas -f one-to-one is suitable for reporting orthologous mappings among all computed assembly to genome mappings.">
+      <option value="map">map - best mapping for long reads</option>
+      <option value="one-to-one">one-to-one - best for mapping orthologous reads</option>
+      <option value="none">None</option>
+    </param>
+  </inputs>
+  <outputs>
+    <data name="mashout" format="paf" label="mashmap on $query.element_identifier" hidden="false"/>
```
Comment (bgruening) [inline tools/mashmap/mashmap.xml:48]:
```suggestion
    <data name="mashout" format="paf"/>
```

File: tools/mashmap/mashmap.xml [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+ 
+ The output is space-delimited with each line consisting of query name, length, 0-based start, end, strand, target name, length, start, end and mapping nucleotide identity.
+ 
+ Map set of query seqences against a list of reference genomes:
+ 
+ mashmap --rl referenceList.txt -q query.fa
+ 
+ File 'referenceList.txt' containing the list of reference genomes should contain path to the reference genomes, one per line.
+ 
+ Source code: https://github.com/marbl/MashMap
```
Comment (bgruening) [inline tools/mashmap/mashmap.xml:81]:
```suggestion
```

File: tools/mashmap/mashmap.xml [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+#end if
+#if len($reflist) == 1:
+ -r '$reflist' -q '$query' && 
+#else
+--rl 'reflist' -q '$query' && 
+#end if
+cp 'mashmap.out'  $mashout]]></configfile>
+  </configfiles>
+  <inputs>
+    <param name="query" type="data" optional="false" label="Query sequences (as fasta) to mash against the references supplied below" help="" format="fasta" multiple="false"/>
```
Comment (bgruening) [inline tools/mashmap/mashmap.xml:32]:
```suggestion
    <param name="query" type="data" format="fasta" label="Query sequences (as FASTA) to mash against the references supplied below"/>
```

File: tools/mashmap/mashmap.bash.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (fubar2) [reply tools/mashmap/mashmap.bash.txt]: done.

File: tools/mashmap/mashmap.xml [hunk 10]
```diff
@@ -0,0 +1,88 @@
+<tool name="mashmap" id="mashmap" version="3.1.3" profile="22.05">
+  <!--Source in git at: https://github.com/fubar2/galaxy_tf_overlay-->
+  <!--Created by toolfactory@galaxy.org at 23/02/2024 21:27:46 using the Galaxy Tool Factory.-->
```
Comment (bgruening) [inline tools/mashmap/mashmap.xml]:
```suggestion
```

File: tools/mashmap/mashmap.xml (hunk 7 above)
Comment (fubar2) [reply tools/mashmap/mashmap.xml:48]: Why not? I like those names!

File: tools/mashmap/mashmap.xml (hunk 3 above)
Comment (fubar2) [reply tools/mashmap/mashmap.xml]: Thanks - I have no idea why those seem to keep disappearing. I am sure I've fixed them all but some git gremlin eats them...

File: tools/mashmap/mashmap.xml (hunk 1 above)
Comment (fubar2) [reply tools/mashmap/mashmap.xml:8]: yes, there is. 
The challenge to automation is that there's never any guarantee that an arbitrary dependency does have a --version command, so the TF uses that compromise...

File: tools/mashmap/LICENSE.txt [hunk 11]
```diff
@@ -0,0 +1,49 @@
+(This is the license file https://github.com/marbl/MashMap/blob/master/LICENSE.txt from the mashmap source repository)
```
Comment (bernt-matthias) [inline tools/mashmap/LICENSE.txt:1]: I don't think we need the mashmap license here.

File: tools/mashmap/mashmap.xml [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+mashmap --pi '$perc_identity' -s '$seqLength' -f '$filtermode' $dense \
+#if int($sketchSize) > 0:
+-J '$sketchSize' \
+#end if
+#if len($reflist) == 1:
+ -r '$reflist' -q '$query' && 
+#else
+--rl 'reflist' -q '$query' && 
+#end if
+cp 'mashmap.out'  '$mashout']]></configfile>
```
Comment (bernt-matthias) [inline tools/mashmap/mashmap.xml:29]: you can just use `from_workdir="mashmap.out"` in the `mashout` output.

File: tools/mashmap/mashmap.xml [hunk 13]
```diff
@@ -0,0 +1,78 @@
+<tool name="mashmap" id="mashmap" version="3.1.3" profile="22.05">
+  <!--Source in git at: https://github.com/fubar2/galaxy_tf_overlay-->
+  <!--Created by toolfactory@galaxy.org at 23/02/2024 21:34:16 using the Galaxy Tool Factory.-->
+  <description>Fast local alignment boundaries</description>
+  <requirements>
+    <requirement version="3.1.3" type="package">mashmap</requirement>
+  </requirements>
+  <version_command><![CDATA[echo "3.1.3"]]></version_command>
+  <command><![CDATA[bash '$runme']]></command>
```
Comment (bernt-matthias) [inline tools/mashmap/mashmap.xml:9]: Why this additional bash call? Galaxy already runs the script that is generated from the `command` section in bash.

Just remove the configfile and move its contents to command.

File: tools/mashmap/mashmap.xml [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+#end if
+#if len($reflist) == 1:
+ -r '$reflist' -q '$query' && 
+#else
+--rl 'reflist' -q '$query' && 
+#end if
+cp 'mashmap.out'  '$mashout']]></configfile>
+  </configfiles>
+  <inputs>
+    <param name="query" type="data" format="fasta" label="Query sequences (as FASTA) to mash against the references supplied below"/>
```
Comment (bernt-matthias) [inline tools/mashmap/mashmap.xml:32]: Is `fasta.gz` possible?

File: tools/mashmap/mashmap.xml [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+    <param name="reflist" type="data" optional="false" label="Reference or references to mash the query sequences on" help="Choose one or more reference sequences to mash the query sequences against." format="fasta" multiple="true"/>
+    <param name="perc_identity" type="float" value="85.0" label="Identity threshold" help="By default, it is set to 85, implying mappings with 85 or more identity should be reported. For example, it can be set to 80to account for more noisy long-read datasets or 95 for mapping human genome assembly to human reference."/>
+    <param name="seqLength" type="integer" value="5000" label="Minimum segment length" help="Default is 5,000 bp. Sequences below this length are ignored. Mashmap provides guarantees on reporting local alignments of length twice this value."/>
```
Comment (bernt-matthias) [inline tools/mashmap/mashmap.xml:35]: Can you  check if it's possible to add `min` and/or `max` to all numeric parameters?

**Comment by fubar2** [comment]
Thanks @bernt-matthias - It works as it is, but in case anyone wants to "fix" it, I'll just leave the code in my fork.

File: tools/mashmap/mashmap.xml (hunk 14 above)
Comment (bgruening) [reply tools/mashmap/mashmap.xml:32]: It seems so, but the results are different ouch!

grafik

I will not include gz inputs for now.

**Comment by fubar2** [comment]
Reopening so @bgruening can transmute this sow's ear into a beautiful silk purse....

**Comment by fubar2** [comment]
Reopening so @bgruening and anyone else interested can transmute this sow's ear into a beautiful silk purse....

**Pull request closed without merging**